### PR TITLE
docker: use pre-built mlir and cva6 toolchain 

### PR DESCRIPTION
**Pull request by jorendumoulin**
this also adds the streamer-gemm and kul-cluster system, until we converge on the single system for all tests

### CODE REVIEW

File: util/container/Dockerfile [hunk 1]
```diff
@@ -225,3 +290,13 @@ COPY --from=snax-streamer-gemm-add-c /src/target/snitch_cluster/sw/snax/ /opt/sn
...
+# COPY --from=snax-kul-cluster-all-narrow /src/sw/snRuntime /opt/snax-kul-cluster-all-narrow/sw/snRuntime
+# COPY --from=snax-kul-cluster-all-narrow /src/target/snitch_cluster/sw/runtime/rtl /opt/snax-kul-cluster-all-narrow/target/snitch_cluster/sw/runtime/rtl
+# COPY --from=snax-kul-cluster-all-narrow /src/target/snitch_cluster/sw/runtime/rtl-generic /opt/snax-kul-cluster-all-narrow/target/snitch_cluster/sw/runtime/rtl-generic
+# COPY --from=snax-kul-cluster-all-narrow /src/target/snitch_cluster/sw/runtime/common /opt/snax-kul-cluster-all-narrow/target/snitch_cluster/sw/runtime/common
+# COPY --from=snax-kul-cluster-all-narrow /src/target/snitch_cluster/sw/snax/ /opt/snax-kul-cluster-all-narrow/target/snitch_cluster/sw/snax
+# COPY --from=snax-kul-cluster-all-narrow /src/sw/math/ /opt/snax-kul-cluster-all-narrow/sw/math/
+# COPY --from=snax-kul-cluster-all-narrow /src/sw/deps/riscv-opcodes /opt/snax-kul-cluster-all-narrow/sw/deps/riscv-opcodes
+# COPY --from=snax-kul-cluster-all-narrow /src/sw/deps/printf /opt/snax-kul-cluster-all-narrow/sw/deps/printf
```
Comment (JosseVanDelm) [inline util/container/Dockerfile:302]: Why is this commented out?

File: util/container/Dockerfile [hunk 2]
```diff
@@ -54,6 +64,15 @@ RUN add-apt-repository -y ppa:deadsnakes/ppa && \
   ln -fs /usr/bin/python3.10 /usr/bin/python3 && \
   ln -fs /bin/python3.10 /bin/python3
 
+# Get prebuilt CVA6 gcc toolchain
+ENV CVA6_GCC_URL=https://github.com/KULeuven-MICAS/llvm-build/releases/download/test-7/custom_toolchain_build.tar.gz
```
Comment (JosseVanDelm) [inline util/container/Dockerfile:68]: test-7 seems like a random identifier?
Also, why do we need to build this ourselves?
There are plenty other sources to get builds from, no? Such as : https://github.com/riscv-collab/riscv-gnu-toolchain/releases

File: util/container/Dockerfile [hunk 3]
```diff
@@ -46,6 +45,17 @@ RUN apt-get update && apt-get install -y lsb-release wget software-properties-co
   # Enable break-system-packages for non-venv package installs
   python3 -m pip config set global.break-system-packages true
 
+# Get prebuilt mlir
+# Set the URL for the download
+ENV MLIR_URL=https://github.com/KULeuven-MICAS/llvm-build/releases/download/test-0/mlir-build.tar.gz
```
Comment (JosseVanDelm) [inline util/container/Dockerfile:50]: test-0?

File: util/container/Dockerfile [hunk 4]
```diff
@@ -182,6 +219,24 @@ RUN git clone https://github.com/kuleuven-micas/snax_cluster /src && \
     make -C target/snitch_cluster bin/snitch_cluster.vlt \
     CFG_OVERRIDE=cfg/snax-streamer-gemm-add-c.hjson -j $(nproc)
 
+FROM base as snax-kul-cluster-all-narrow
+
+# RUN git clone https://github.com/kuleuven-micas/snax_cluster /src && \
```
Comment (JosseVanDelm) [inline util/container/Dockerfile:224]: Commented code

**Review by JosseVanDelm — COMMENTED** [review]
I overall like this, but I'm unsure about the current versioning strategy and I don't feel like we should build gcc ourselves (do we need to?)

**Comment by jorendumoulin** [comment]
this is a draft @JosseVanDelm 
that explains the test tags
CVA6 official repo says this:
image

the commented code is because the container was running out of memory trying to build the system, because of many tcdm connections. not the case anymore since #254

**Comment by jorendumoulin** [comment]
Now open for review!

File: util/container/Dockerfile [hunk 5]
```diff
@@ -46,6 +45,17 @@ RUN apt-get update && apt-get install -y lsb-release wget software-properties-co
   # Enable break-system-packages for non-venv package installs
   python3 -m pip config set global.break-system-packages true
 
+# Get prebuilt mlir
+# Set the URL for the download
+ENV MLIR_URL=https://github.com/KULeuven-MICAS/llvm-build/releases/latest/download/mlir.tar.gz
```
Comment (JosseVanDelm) [inline util/container/Dockerfile:50]: Can we pin the version here instead as well?
```
ENV MLIR_URL=https://github.com/KULeuven-MICAS/llvm-build/releases/ <version> /download/mlir.tar.gz
```

**Review by JosseVanDelm — APPROVED** [review]
Thank you so much @jorendumoulin !!!

I like this, but can we pin the versions in the llvm-build repo here as well maybe?

**Review by rgantonio — APPROVED** [review]
OK! Nice. So we can use CVA6 with this? 😄

**Comment by JosseVanDelm** [comment]
@rgantonio from what I understand that is the current goal yes.
Ideally we would use just clang for everything, but @IveanEx has tried this and there were some issues with using the linker scripts and libraries and such (LLD is likely to never link it with the linker scripts of occamy, because it was only tested with GNU LD and it contains some GNU-specific LLVM-wontfix code - https://fosdem.org/2024/events/attachments/fosdem-2024-2340-linker-scripts-in-lld-and-how-they-compare-with-gnu-ld/slides/22752/Fosdem_linker_scripts_2024_uG7XLOI.pdf ). 
Completely getting this working would likely require another cowboy effort such as #57 
Since we can also just link gcc and clang binaries together, we just have both of them here.

**Comment by IveanEx** [comment]
When doing some initial attempts on compiling the sw, some weird errors are raised (at linking stage). The simpler software, e.g., bootrom, seems to have no problems. Now @Konste11ation is investigating it, and please do not merge after we solved all the problems.

**Comment by IveanEx** [comment]
> @rgantonio from what I understand that is the current goal yes. Ideally we would use just clang for everything, but @IveanEx has tried this and there were some issues with using the linker scripts and libraries and such (LLD is likely to never link it with the linker scripts of occamy, because it was only tested with GNU LD and it contains some GNU-specific LLVM-wontfix code - https://fosdem.org/2024/events/attachments/fosdem-2024-2340-linker-scripts-in-lld-and-how-they-compare-with-gnu-ld/slides/22752/Fosdem_linker_scripts_2024_uG7XLOI.pdf ). Completely getting this working would likely require another cowboy effort such as #57 Since we can also just link gcc and clang binaries together, we just have both of them here.

Maybe another solution is to establish a new runtime and compilation flow for our system...

**Review by IveanEx — APPROVED** [review]
HeMAiA test passed! Thanks @Konste11ation !

**Pull request merged**